File: scala3/src/main/scala/io/github/sps23/interview/preparation/async/ApiResponse.scala
package io.github.sps23.interview.preparation.async

/** Represents an API response with data from an external service.
  *
  * This case class demonstrates Scala's approach to immutable data in async contexts. Case classes
  * work seamlessly with Scala Futures for representing results from async operations.
  *
  * @param source
  *   the name of the API source
  * @param data
  *   the response data
  * @param timestamp
  *   when the response was received
  */
case class ApiResponse(source: String, data: String, timestamp: Long):
  require(source != null, "Source cannot be null")
  require(data != null, "Data cannot be null")
  require(timestamp > 0, "Timestamp must be positive")

object ApiResponse:
  /** Creates an ApiResponse with the current timestamp.
    *
    * @param source
    *   the API source name
    * @param data
    *   the response data
    * @return
    *   new ApiResponse instance
    */
  def apply(source: String, data: String): ApiResponse =
    ApiResponse(source, data, System.currentTimeMillis())

File: kotlin/src/main/kotlin/io/github/sps23/interview/preparation/async/ApiResponse.kt
package io.github.sps23.interview.preparation.async

/**
 * Represents an API response with data from an external service.
 *
 * This data class demonstrates Kotlin's approach to immutable data in async contexts.
 * Data classes work seamlessly with Kotlin Coroutines for representing results from
 * async operations.
 *
 * @property source the name of the API source
 * @property data the response data
 * @property timestamp when the response was received
 */
data class ApiResponse(
    val source: String,
    val data: String,
    val timestamp: Long = System.currentTimeMillis(),
) {
    init {
        require(source.isNotBlank()) { "Source cannot be blank" }
        require(data.isNotBlank()) { "Data cannot be blank" }
        require(timestamp > 0) { "Timestamp must be positive" }
    }

    companion object {
        /**
         * Creates an ApiResponse with the current timestamp.
         *
         * @param source the API source name
         * @param data the response data
         * @return new ApiResponse instance
         */
        fun of(
            source: String,
            data: String,
        ): ApiResponse = ApiResponse(source, data)
    }
}
